### PR TITLE
Trait concepts

### PR DESCRIPTION
**Pull request by Ephraim-nonso**

### CODE REVIEW

**Comment by 0xNeshi** [comment]
I will check out the PR tomorrow, I've been a bit busy

File: concepts/traits/about.md [hunk 1]
```diff
@@ -1 +1,315 @@
...
+    println!("Circle area: {}", ShapeGeometry::area(circ)); //78
+    println!("Circle boundary: {}", ShapeGeometry::boundary(circ)); //31
+}
+```
+
+The implementation visibility for both the `CircleGeometry` and `RectangleGeometry` does not need to be public since the `ShapeGeometry` trait is made public and should be easily found by the compiler.
+
+## Impl Aliases
+
+Implementation, when imported, can be labeled as `impl` as an alternative name. This helps when instantiating a generic implementation for concrete types. Let's define a trivial `Two` implementation for all types that implement the `One` trait by simply adding twice the value of one and returning the summation. We can afterwards make available for use an implementation that works for u8 and u128 types.
```
Comment (0xNeshi) [inline concepts/traits/about.md:233]:
```suggestion
Implementations can be aliased when imported, using the `impl` keyword. This helps when instantiating a generic implementation for concrete types. Let's define a trivial implementation of trait `Two` for all types that implement the `One` trait by simply adding twice the value of `One` and returning the summation. We can afterwards expose the `Two` implementation that works only for `u8` and `u128` types.
```
nit: wording

File: concepts/traits/about.md [hunk 2]
```diff
@@ -1 +1,315 @@
...
+    }
+}
+
+pub impl U8Two = one_based::TwoImpl<u8>;
+pub impl U128Two = one_based::TwoImpl<u128>;
+```
+
+The generic implementation can be written in a private module and with the use of impl alias, we can instantiate the generic implementation of these two concrete types. The benefit of using a private and unexposed generic implementation is that it prevents code duplication across files. This as well keeps the public API simple and clean.
+
+## Negative Implementation
```
Comment (0xNeshi) [inline concepts/traits/about.md:256]: Let's describe only the production-ready features of Cairo, not experimental ones. 
**Remove this section for now**, and we can add it back once Negative Impls are part of the core language. 

Created an issue to track this https://github.com/exercism/cairo/issues/143

File: concepts/traits/introduction.md [hunk 3]
```diff
@@ -1 +1,3 @@
 # Introduction
+
+In Cairo, a trait is a series of methods that can be implemented by a data type. When a trait is implemented for a type, the methods defined in that trait can be called on instances of that type. A trait is not limited to only a singular type but when combined with generic type, the functionality defined in the trait can be invoked on a particular type and can share these methods with other data types.
```
Comment (0xNeshi) [inline concepts/traits/introduction.md:3]:
```suggestion
In Cairo, a trait is a series of methods that can be implemented by a data type. When a trait is implemented for a type, the methods defined in that trait can be called on instances of that type. A trait can be defined with generic types, defining behavior shared by multiple data types.
```
nit: wording

**Comment by 0xNeshi** [comment]
@Ephraim-nonso need any help? Let me know if you need me to further clarify some of my comments

**Comment by Ephraim-nonso** [comment]
Hi. I'm sorry I had an issue with my working device. I've it all fixed now and will make the changes according to the comment this coming week.

File: concepts/traits/about.md [hunk 4]
```diff
@@ -144,23 +144,22 @@ fn main() {
...
-The default implementation returns `"(Read more...)"` ByteArray when this trait is implemented on any type. To use this default, you declare the impl block with an empty block.
 
-```rust
-impl NewsArticleSummary of Summary<NewsArticle> {}
+    let tweet = Tweet {
+        username: "Alice",
+        content: "Cairo is fun and engaging.",
+        reply: true,
+        retweet: true,
+    }
```
Comment (0xNeshi) [inline concepts/traits/about.md:155]: The example does not compile due to the missing semicolon `;` here and the missing `Tweet` import. Add it and format the code.

File: concepts/traits/about.md [hunk 5]
```diff
@@ -1 +1,253 @@
...
+
+> **Note:** Traits share similar feature to what is known as interfaces in other programming languages with some differences.
+
+## Trait definition
+
+Defining a trait helps identify the function signature that would make up the set of behaviors a type can implement in order to achieve a result. These methods form the behavior of a type. If we can call the same methods on different types, this implies that the involved types have same behavior.
+
+```rust
+#[derive(Drop)]
+    pub struct NewsArticle {
```
Comment (0xNeshi) [inline concepts/traits/about.md:13]: nit: format the code

File: concepts/traits/about.md [hunk 6]
```diff
@@ -1 +1,253 @@
...
+        author: "Cairo Digger",
+        content: "Cairo is a new programming language for zero-knowledge proofs",
+    };
+
+    let tweet = Tweet {
+        username: "EliBenSasson",
+        content: "Crypto is full of short-term maximizing projects. \n @Starknet and @StarkWareLtd are about long-term vision maximization.",
+        reply: false,
+        retweet: false
+    }; // Tweet instantiation
```
Comment (0xNeshi) [inline concepts/traits/about.md:93]:
```suggestion
    };
```
nit

**Review by 0xNeshi — CHANGES_REQUESTED** [review]
Just a few more small changes and we can merge

**Review by 0xNeshi — APPROVED** [review]
Very good work!

**Comment by Ephraim-nonso** [comment]
Thank you. Apologies for the delayed work pace. I'd pick another.

**Comment by 0xNeshi** [comment]
@Ephraim-nonso no problem, feel free to pick any issue you like

**Pull request merged**